### ISSUE TITLE
Just migrated my UWP app to Sentry from AppCenter but the release build is crashing on startup. Debug build runs fine

### ISSUE THREAD

**Issue by dpaulino** [getsentry/sentry-dotnet#3240]
### Package

Sentry

### .NET Flavor

.NET

### .NET Version

UWP

### OS

Windows

### SDK Version

4.2.1

### Steps to Reproduce

1. Get ambie white noise's code: https://github.com/jenius-apps/ambie 
2. Edit appsettings.resw > TelemetryApiKey. Add your sentry DSN in that field
3. Deploy debug build. Observe it runs fine
4. But release build crashes

### Expected Result

No crashes on startup

### Actual Result

Crashes on startup

**Comment by bruno-garcia** [comment]
> I'm unblocked but it's not ideal. I want to maximize performance using trimming. AppCenter didn't have issues like this, so I was hoping Sentry would be the same.

yeah checking if u're unblocked as that's the first step. [We made the whole code base AOT friend for Native AOT as I mentioend on twitter](https://sentry.engineering/blog/should-you-could-you-aot). But regarding UWP I'm not sure how much that was tested. Sentry uses only Mac's so Windows is mainly tested in CI and @jamescrosswell, @lucas-zimerman and @vaind use Windows too but are not building UWP apps.

So we'll need a bit of your support figuring out what we need here to sort out. It sounds like it's including the `rd.xml` with those rules as part of the SDK, does that sound right?

**Comment by jamescrosswell** [comment]
@dpaulino maybe you can help me out. I'm having difficulty reproducing this. I basically ran through the steps you outlined:
1. Cloned ambie
2. Set the TelemetryApiKey
3. Ran it in Debug (all good)
4. Ran it in Release mode (again, all good)

What command are you running to build a release version of the application (and so what should I run here to reproduce the issue)?

**Comment by tipa** [comment]
The problem can be reproduce as follows:
- make new UWP app from template in VS22
- comment out this line in the Default.rd.xml file: `<Assembly Name="*Application*" Dynamic="Required All" />`
- add Sentry nuget and call Init, set a valid DSN: `Sentry.SentrySdk.Init(o => o.Dsn = "https://....@.....ingest.sentry.io/....");`
- run in either Release or Debug with csproj property `UseDotNetNativeToolchain` set to `true`
Here's an example project: [sentry_test.zip](https://github.com/getsentry/sentry-dotnet/files/14863466/sentry_test.zip)

**Comment by vaind** [comment]
May be related to the use of reflection: https://support.syncfusion.com/kb/article/7557/how-to-overcome-the-crash-system-reflection-missingmetadataexception-when

**Comment by tipa** [comment]
Yes, it absolutely has to do with reflection.
Maybe that helps making the app trimming-safe:
https://devblogs.microsoft.com/ifdef-windows/leveraging-trimming-to-make-the-microsoft-store-faster-and-reduce-its-binary-size/
https://devblogs.microsoft.com/dotnet/net-native-deep-dive-making-your-library-great/

**Comment by bruno-garcia** [comment]
Yeah if it's just about keeping those few lines it's OK. @jamescrosswell is this something you can help with a fix?
I wonder if we can get a sample app built in release mode (something that today would repro the crash) and launch in CI to make sure it's working. To avoid regression. We have such a smoketest with [IL2CPP on the Sentry Unity SDK](https://github.com/getsentry/sentry-unity/) which @vaind setup

**Comment by tipa** [comment]
Those lines would at least prevent the crash at SDK initialization. I don't know if there's other classes that are trimmed away and would cause crashes when they are attempted to be instantiated through reflection later on (e.g. when an event occurs or breadcrumbs are collected)...

**Comment by bitsandfoxes** [comment]
There was so much effort put into making the SDK compliant with AOT and Trimming, let's push it over the edge.

**Comment by lucas-zimerman** [comment]
> There was so much effort put into making the SDK compliant with AOT and Trimming, let's push it over the edge.

We may need to add some extra tests for .NET Native

also a sample event from a release build: https://sentry-sdks.sentry.io/issues/5158519505/events/f681faaa5d844687827d66cb70be3ea9/?project=5428537

(using the main code from (https://github.com/jenius-apps/ambie)

But I was able to reproduce it with the [sentry_test.zip](https://github.com/getsentry/sentry-dotnet/files/14863466/sentry_test.zip) in release

**Comment by jamescrosswell** [comment]
> Yeah if it's just about keeping those few lines it's OK. @jamescrosswell is this something you can help with a fix?

I can definitely help with this... just trying to balance efforts with some feature work that is supposed to go GA at the end of this week as well - so a prioritisation balancing act. 

Generally we should get compiler warnings for dynamically generated code... unless we've incorrectly suppressed those warnings. 

There are two ways to react to those warnings - one is to tell the compiler which types to keep when trimming (@lucas-zimerman I see that's the approach you've taken above). The other is to rewrite the code so that it doesn't depend on things like reflection... typically using source generators.

@lucas-zimerman do you have the bandwidth to look into a fix for this or do you want/need me to pick it up?

**Comment by dpaulino** [comment]
> @dpaulino maybe you can help me out. I'm having difficulty reproducing this. I basically ran through the steps you outlined:
> 
> 1. Cloned ambie
> 
> 2. Set the TelemetryApiKey
> 
> 3. Ran it in Debug (all good)
> 
> 4. Ran it in Release mode (again, all good)
> 
> 
> 
> What command are you running to build a release version of the application (and so what should I run here to reproduce the issue)?

You'll want to check out this commit in order to reproduce the issue: 1fdeb439ec22008a5bac0e9083a801e75cade75c. 

I removed sentry from the latest main branch commits, and so that's why you're not able to repro. 

I might bring sentry back after re-evaluating the trimming support later.

**Comment by jamescrosswell** [comment]
> You'll want to check out this commit in order to reproduce the issue: 1fdeb439ec22008a5bac0e9083a801e75cade75c.

Cool thanks @dpaulino - it looks like we can reproduce this now. Will track progress here on this ticket.

**Comment by lucas-zimerman** [comment]
@jamescrosswell  I have time for doing this but not the integration test with UWP.
Is there any preference on which fix should I apply to Sentry .NET?

**Comment by jamescrosswell** [comment]
> Is there any preference on which fix should I apply to Sentry .NET?

Awesome, thanks @lucas-zimerman 

Personally I'd favour an approach that removes the reliance on reflection... mainly because it gives us more static compiler checks moving forward. But that's not particularly easy in this case... we can't add a `ISdkIntegration.IntegrationName` property or anything to that interface (breaking change) so it'd require adding an additional internal interface and fiddling around - all just for one log statement... which seems like overkill. Given that, I think the solution you've come up with is probably the most practical.

It would be good to work out why we're not getting any compiler warnings/errors for `GetType().Name`. We've enabled [AOT Compatibility Analyzers](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net8plus%2Cwindows#aot-compatibility-analyzers):
https://github.com/getsentry/sentry-dotnet/blob/4b33c36a22aec55d3e1c75eca51acc679a1e4d21/src/Sentry/Sentry.csproj#L28

I would expect some kind of warning for code like this that isn't AOT compatible then. Without those warnings, it's really hard to have any confidence that there aren't hundreds of other little problems like this sprinkled through our code.

**Comment by bruno-garcia** [comment]
> It would be good to work out why we're not getting any compiler warnings/errors for GetType().Name. We've enabled [AOT Compatibility Analyzers](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net8plus%2Cwindows#aot-compatibility-analyzers):

My guess is that's only for Native AOT, .NET 6+ etc. Where UWP is .NET Native (the Windows-only one that's been around for a lot longer). 

UWP compiled and ran (without trimming at least) for a long time and we had no warnings in CI. So I wouldn't bet only on these compatibility analyzer, we need something in CI, at least a smoke test

**Comment by tipa** [comment]
I believe I have found another problem when trimming is enabled. I've noted that the stack traces I see for crashes in my UWP apps have a significantly shorter stack trace with less  information compared to what I've seen in the AppCenter SDK.
I've did some tests with this project ([SentryTest.zip](https://github.com/getsentry/sentry-dotnet/files/14937380/SentryTest.zip)) and noted that when trimming is enabled (the `Assembly` line in `Default.rd.xml` is commented out) I get this stack trace:

```
System.Exception: trimming
  Module "SentryTest.MainPage", in ComboBox_SelectionChanged
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "System.Action`1", in Invoke
```

while when there is no trimming enabled I get this stack trace:
```
System.Exception: no trimming
  Module "SentryTest.MainPage", in ComboBox_SelectionChanged
  Module "SentryTest.MainPage", in MainPage_Loaded
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "System.Action`1", in Invoke
```

Both times I built the app in Release mode (.NET Native)

**Comment by tipa** [comment]
I verified that the stack trace is available - when I register the `UnhandledException` handler ([as per the docs](https://docs.sentry.io/platforms/dotnet/guides/winui/#configuration-with-trimming)) and logging the exception by other ways, I can see that there is much, much more information in the stack trace than what is shown on the Sentry dashboard. For some exceptions, all I get is a stack trace like this:
```
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "System.Action`1", in Invoke
```
Should I file a separate issue for this?

**Comment by bruno-garcia** [comment]
> I verified that the stack trace is available - when I register the `UnhandledException` handler ([as per the docs](https://docs.sentry.io/platforms/dotnet/guides/winui/#configuration-with-trimming)) and logging the exception by other ways, I can see that there is much, much more information in the stack trace than what is shown on the Sentry dashboard. For some exceptions, all I get is a stack trace like this:
> 
> ```
>   Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
>   Module "System.Action`1", in Invoke
> ```
> 
> Should I file a separate issue for this?

By default, Sentry filters the frames to show only application code. You can click a button to show all frames. Could that be it? Or is that the raw stack trace you copied here? 

Could you please link to the even in Sentry?

**Comment by tipa** [comment]
Today I received another funny stack trace (it is the raw stack trace) - https://timopartl.sentry.io/issues/5216095235:
```
System.ArgumentException: Value does not fall within the expected range.
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "System.Runtime.CompilerServices.TaskAwaiter", in GetResult
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "System.Runtime.CompilerServices.TaskAwaiter", in GetResult
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "System.Runtime.CompilerServices.TaskAwaiter", in GetResult
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "System.Runtime.CompilerServices.TaskAwaiter", in GetResult
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "System.Runtime.CompilerServices.TaskAwaiter", in GetResult
```

It's from a manually captured exception, but the crash report I was referring to in my last comment was from a crash (and also it was the raw stack trace), here's the link: https://timopartl.sentry.io/issues/5178256155


Here's another example project to reproduce the issue. I log the stack trace manually as breadcrumb, and this is what I see in the dashboard: [Sentry_UWP.zip](https://github.com/getsentry/sentry-dotnet/files/15022198/Sentry_UWP.zip)


stack trace:
![image](https://github.com/getsentry/sentry-dotnet/assets/2863198/2a70a4df-a443-4039-94e5-f540e498b107)

stack trace from breadcrumb:
```
at Sentry_UWP.SubPage.Toggle_Toggled(Object, RoutedEventArgs) + 0x8
   at Windows.Foundation.AsyncOperationProgressHandler`2.Invoke(IAsyncOperationWithProgress`2, TProgress) + 0x2e
   at __Interop.Intrinsics.HasThisCall__13[TArg0](Object, IntPtr, Object, TArg0) + 0x36
   at Sentry_UWP!<BaseAddress>+0x4d29d0
   at System.Runtime.InteropServices.McgMarshal.ThrowOnExternalCallFailed(Int32, RuntimeTypeHandle) + 0x21
   at __Interop.ComCallHelpers.Call(__ComObject, RuntimeTypeHandle, Int32, SByte) + 0xc0
   at __Interop.ForwardComStubs.Stub_29[TThis](__ComObject, Boolean, Int32) + 0x37
   at Sentry_UWP.SubPage.OnNavigatedTo(NavigationEventArgs) + 0x56
   at Windows.UI.Xaml.Controls.Page.global::Windows.UI.Xaml.Controls.IPageOverrides.OnNavigatedTo(NavigationEventArgs) + 0xa
   at Sentry_UWP!<BaseAddress>+0x4e498e
```

This is a very simple example so there's not a lot information missing, but I would at least expect to see the line
`Sentry_UWP.SubPage.OnNavigatedTo(NavigationEventArgs) `. If you want me to, I can try to produce a crash that has a longer stack trace with more information missing. But I think even this small exmaple shows that there's something wrong.

**Comment by bruno-garcia** [comment]
Yeah definitely something seems wrong. If we are seeing different frames between what we can build in the Sentry stack trace protocol vs a stringified version, that's odd. 

We rely on .NET to give us the stack trace, e.g:

https://github.com/getsentry/sentry-dotnet/blob/233e95b5a2104819c735a29490c1aaf6b336f1fd/src/Sentry/Extensibility/SentryStackTraceFactory.cs#L33

Overtime with Ben.Demystifier, server-side symbolication, AOT things got a bit more complicated 

https://github.com/getsentry/sentry-dotnet/blob/233e95b5a2104819c735a29490c1aaf6b336f1fd/src/Sentry/Internal/DebugStackTrace.cs#L179-L181

https://github.com/getsentry/sentry-dotnet/blob/233e95b5a2104819c735a29490c1aaf6b336f1fd/src/Sentry/Internal/DebugStackTrace.cs#L247-L250

So maybe a bug was introduced somewhere. @jamescrosswell @bitsandfoxes seems to only affect .NET Native but @tipa has a zip with a repro a links to errors as examples. I'd say this is p0 but that's more up to @stephanie-anderson I imagine.

**Comment by jamescrosswell** [comment]
I think we've strayed away from the original problem that @dpaulino reported and that @lucas-zimerman [suggested a fix for](https://github.com/getsentry/sentry-dotnet/issues/3240#issuecomment-2043120591), so I created a separate issue to look into the stack traces:
- https://github.com/getsentry/sentry-dotnet/issues/3316

**Comment by jamescrosswell** [comment]
> My guess is that's only for Native AOT, .NET 6+ etc. Where UWP is .NET Native (the Windows-only one that's been around for a lot longer).

That could definitely be an issue when using UWP with `UseDotNetNativeToolchain` on .NET 7 or earlier. 

Our checks relating to AOT only apply to .NET 8 and later:
https://github.com/getsentry/sentry-dotnet/blob/93b8508d76294cd8c815aa0831cd04fa7d932518/src/Directory.Build.props#L38-L46

And even on .NET 8, we never tested to see if this logic also works with Trimming in UWP:
https://github.com/getsentry/sentry-dotnet/blob/2e4649c3168f8bd574348a14c68ef19c85bd23e0/src/Sentry/Internal/AotHelper.cs#L16-L20

**Comment by jamescrosswell** [comment]
So this definitely relates to trimming. Many of the fixes that we put in place to enable trimming for AOT rely on functionality that's only available in .NET 8 (source generators for various things to avoid the use of reflection)... None of that code is being executed in the context of a UWP application that targets .NET Framework, so when `UseDotNetNativeToolchain` is set for UWP apps, the release builds for these aren't working. 

Even if we get these apps past the "launching" phase then (so SentrySdk.Init itself doesn't fail) I think we're likely to have lots of other problems when these applications are running. 

I can't think of any good solution for this (nor can I think how this might have worked previously - the SentrySDK has made use of reflection in various different places for quite some time) 🤔

**Comment by tipa** [comment]
It is quite possible that it also didn't work in the past - by default UWP did not use trimming and I expect only the small subset of developers who opted-in and put in the effort to make their app trimming-safe would encounter this problem. With AppCenter being scheduled for retirement, some UWP developers (like @dpaulino and me) were looking for an alternative and I assume that's why the problem only surfaced now.
If I'm not completely wrong, source generators can be used in UWP as well (I am using the `System.Net.Http.Json` package, that does a lot of source generation, in UWP)

**Comment by jamescrosswell** [comment]
> If I'm not completely wrong, source generators can be used in UWP as well (I am using the `System.Net.Http.Json` package, that does a lot of source generation, in UWP)

The source generators can yes (they're a design time thing). However the resulting classes generated by the source generators we used to enable AOT in Sentry rely on features of .NET 8 (which is why we only support AOT on .NET 8 and above). 

In the short to medium term, I think we need some kind of warning if you to initialize Sentry with trimming enabled when targeting net7.0 or below (as the SDK doesn't currently support that).

Hopefully we can carve out some time eventually to enable native compilation for .NET Framework applications... but that's likely a significant amount of work and so I can't make any promises on the timing for that at the moment.

**Comment by bruno-garcia** [comment]
> native compilation for .NET Framework applications.

wdyt? Don't think qe need to invest in .NET Framework from that pov. UWP isn't .NET Framework btw, it's .NET Native.

**Comment by dpaulino** [comment]
Any updates on whether or not trimming will be supported in uwp? One of my apps is using sentry, but I migrated away from sentry in my second app because of the lack of trimming support. I want to use sentry again for the crash analytics, but this issue is blocking me.

**Comment by jamescrosswell** [comment]
> Any updates on whether or not trimming will be supported in uwp?

@bitsandfoxes this sounds like a priorities call (in your court).

My 2c is that we should invest more in ensuring AOT/trimming is robust for ASP.NET Core, iOS and Android before we try to add support for other platforms, so UWP support is probably further down the line. We're pretty stretched so it's hard to see us doing both of these things.

**Comment by bitsandfoxes** [comment]
> we should invest more in ensuring AOT/trimming is robust for ASP.NET Core, iOS and Android

I heavily agree. 

We're tracking efforts to increase our support for UWP with https://github.com/getsentry/sentry-dotnet/issues/3347 and as far as I can tell this would fall in there as well?

**Comment by jamescrosswell** [comment]
> We're tracking efforts to increase our support for UWP with [#3347](https://github.com/getsentry/sentry-dotnet/issues/3347) and as far as I can tell this would fall in there as well?

I think a UWP package would be a pre-requisite for supporting trimming (see [this comment](https://github.com/getsentry/sentry-dotnet/issues/3561#issuecomment-2313969467)).

**Comment by bruno-garcia** [comment]
@dpaulino do you know what would it take to support it?

AOT is working OK for .NET (Core) but UWP isnt' something we're historically focused on although we have folks using it. Not sure in what cases it fails.

**Comment by bruno-garcia** [comment]
> As of October 2021, Microsoft officially deprecated UWP

Unfortunately UWP is in life support by Microsoft. So in the next major release of this SDK, we'll be deprecating UWP support.
